storage/badger/operation/reader_batch_writer.go

```diff
@@ -0,0 +1,129 @@
+package operation
+
+import (
+	"errors"
+	"io"
+	"sync"
+
+	"github.com/dgraph-io/badger/v2"
+
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/storage"
+)
+
+type ReaderBatchWriter struct {
+	db    *badger.DB
+	batch *badger.WriteBatch
+
+	addingCallback sync.Mutex // protect callbacks
+	callbacks      []func(error)
+}
+
+var _ storage.BadgerReaderBatchWriter = (*ReaderBatchWriter)(nil)
+
+func (b *ReaderBatchWriter) GlobalReader() storage.Reader {
```

Comment:
Before merging, please add documentation for these and other public methods. I wrote a version in this comment for the reader/writer methods: https://github.com/onflow/flow-go/pull/6381#discussion_r1727719819.